compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp

```diff
@@ -367,38 +446,70 @@ class ProcessTransferWrite final
-                                 adaptor.getIndices(), rewriter, loc);
+    FailureOr<SmallVector<Value>> indices =
+        adjustIndices(scalarMemrefType, vectorMemrefType, adaptor.getIndices(),
+                      rewriter, loc);
     if (failed(indices))
-      return failure();
+      return rewriter.notifyMatchFailure(write, "failed to adjust indices");
 
     // If the transfer_write can be replaced by a store after vectorization cast
     // the original value and use StoreOp.
     if (*vectorMemrefElemSize == *writeVecSize) {
-      Type serializedVecType = VectorType::get(
-          writeVectorType.getNumElements(), writeVectorType.getElementType());
-      Value data = rewriter.create<vector::ShapeCastOp>(loc, serializedVecType,
-                                                        adaptor.getVector());
-      data = rewriter.create<vector::BitCastOp>(
-          loc, vectorMemrefType.getElementType(), data);
+      Value data = rewriter.create<vector::BitCastOp>(loc, memrefVectorType,
+                                                      adaptor.getVector());
       rewriter.replaceOpWithNewOp<memref::StoreOp>(
           write, data, adaptor.getSource(), indices.value());
-    } else {
-      rewriter.replaceOpWithNewOp<vector::TransferWriteOp>(
-          write, adaptor.getVector(), adaptor.getSource(), indices.value());
+      return success();
     }
+
+    // We use the minimal vector seen in all transfer uses as the vectorized
+    // memref element type, so it's guaranteed that the read vector size is
+    // some multiple of that.
+    assert(*writeVecSize % *vectorMemrefElemSize == 0);
+
+    MLIRContext *context = rewriter.getContext();
+    AffineExpr sym0, sym1;
+    bindSymbols(context, sym0, sym1);
+    auto addMap = AffineMap::get(0, 2, {sym0 + sym1}, context);
+
+    // We have checked that this transfer op has identity map, so we increase
+    // the innermost index to perform continuous reads.
+    Value oldIndex = indices->back();
+
+    int vectorCount = *writeVecSize / *vectorMemrefElemSize;
+
+    SmallVector<int64_t> offsets(writeVectorType.getRank(), 0);
+    SmallVector<int64_t> sizes(writeVectorType.getRank(), 1);
+    SmallVector<int64_t> strides(writeVectorType.getRank(), 1);
+    sizes.back() = memrefVectorType.getNumElements();
+
+    for (int i = 0; i < vectorCount; ++i) {
+      offsets.back() = i * memrefVectorType.getNumElements();
+      auto slice = rewriter.create<vector::ExtractStridedSliceOp>(
+          loc, adaptor.getVector(), offsets, sizes, strides);
+      auto component =
+          rewriter.create<vector::BitCastOp>(loc, memrefVectorType, slice);
+      Value iVal = rewriter.create<arith::ConstantIndexOp>(loc, i);
+      indices->back() = rewriter.create<affine::AffineApplyOp>(
+          loc, addMap, ValueRange{oldIndex, iVal});
+      rewriter.create<memref::StoreOp>(loc, component, adaptor.getSource(),
+                                       *indices);
```

Comment:
Do we have some preference when it comes to `vector` vs. `memref` `load`/`store`? Just curious.

Reply:
`vector.load` performs vectorized load from a scalar-element-type memref. We don't have a direct correspondance in SPIR-V _in logical addressing mode_--In SPIR-V scalar-element-type memref translates to scalar-element-type runtime array. Access chains to it gives us pointers to scalar values; there is no way to get a pointer to a vector of arbiarty size there. OpBitcast the pointer might seems a way; but it's explicitly forbidden in logical addressing mode by excluded in the allowed op list in "2.16.1. Universal Validation Rules". (It might be actually supported by some drivers; though still ill semantics territory.) So that's why we have this pass entirely--trying to change the memref element type to vector type so we can use the basic `memref.load` to load vectors. It's a pain. Physical addressing mode makes this easier as we can do whatever with the pointers.